File: MultipleLinearRegression.py
#Multiple Linear Regression

#import libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#import dataset
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].values
Y = dataset.iloc[:, 4].values


#encode the categorical data
#encode independent variable "state"
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:, 3] = labelencoder_X.fit_transform(X[:, 3])
onehotencoder = OneHotEncoder(categorical_features = [3])
X = onehotencoder.fit_transform(X).toarray()

#avoiding dummy variable trap
X = X[:, 1:]

#splitting the dataset into train and test
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size = 1/5, random_state = 0)

#fitting Multiple Linear Regression to the Training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

#predicting the test set results
y_pred = regressor.predict(X_test)

#building the optimal model using backward elimination
import statsmodels.formula.api as sm
#add a column of ones to matrix of features
    #make X0 = 1 and help make the model
X = np.append(arr = np.ones((50, 1)). astype(int), values = X, axis = 1)
#get a matrix of optimal features
#select a significance level
    #SL = 0.05
#OLS = ordinary least squares
X_opt = X[:, [0,1,2,3,4,5]]
regressor_OLS = sm.OLS(endog = Y, exog = X_opt).fit()
regressor_OLS.summary()
X_opt = X[:, [0,1,3,4,5]]
regressor_OLS = sm.OLS(endog = Y, exog = X_opt).fit()
regressor_OLS.summary()
X_opt = X[:, [0,3,4,5]]
regressor_OLS = sm.OLS(endog = Y, exog = X_opt).fit()
regressor_OLS.summary()
X_opt = X[:, [0,3,5]]
regressor_OLS = sm.OLS(endog = Y, exog = X_opt).fit()
regressor_OLS.summary()
X_opt = X[:, [0,3]]
regressor_OLS = sm.OLS(endog = Y, exog = X_opt).fit()
regressor_OLS.summary()
#finished and obtained the optimal x

File: MultipleLinearRegression.r
#Multiple Linear Regression

#import dataset
dataset <- read.csv("50_Startups.csv")
sapply(dataset, "class")

#encoding the categorical variable, state, to improve model
dataset$State = factor(dataset$State,
                         levels = c('New York', 'California', 'Florida'),
                         labels = c(1, 2, 3))

#split data set (training and testing)
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == T)
test_set = subset(dataset, split == F)

#Fitting Multiple Linear Regression to the training set
regressor <- lm(formula = Profit ~., data = training_set)
summary(regressor)
#using only highly significant features
regressor1 <- lm(formula = Profit ~ R.D.Spend, data = training_set)
summary(regressor)

#Predicting the test set results
y_pred = predict(regressor, newdata = test_set)
View(y_pred)
y_pred

y_pred1 = predict(regressor1, newdata = test_set)
View(y_pred1)
y_pred1

#building the optimal model using backward elimination significance level of 7% for each feature
regressor <- lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
                data = dataset)
summary(regressor)
regressor <- lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
                data = dataset)
summary(regressor)
regressor <- lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
                data = dataset)
summary(regressor)
